pkg/controllers/tagging/tagging_controller.go

```diff
@@ -135,19 +143,34 @@ func (tc *Controller) process() bool {
 		workItem, ok := obj.(*workItem)
 		if !ok {
 			tc.workqueue.Forget(obj)
-			utilruntime.HandleError(fmt.Errorf("expected workItem in workqueue but got %#v", obj))
+			err := fmt.Errorf("expected workItem in workqueue but got %#v", obj)
+			utilruntime.HandleError(err)
+			recordWorkItemErrorMetrics("dequeue_work_item")
 			return nil
 		}
 
+		timeTaken := time.Since(workItem.enqueueTime).Seconds()
+		recordWorkItemLatencyMetrics("dequeue_work_item", timeTaken)
+
 		err := workItem.action(workItem.node)
 		if err != nil {
-			// Put the item back on the workqueue to handle any transient errors.
-			tc.workqueue.AddRateLimited(workItem)
-			return fmt.Errorf("error finishing work item '%v': %s, requeuing", workItem, err.Error())
+			if workItem.requeuingCount < maxRequeuingCount {
+				// Put the item back on the workqueue to handle any transient errors.
+				workItem.requeuingCount++
+				tc.workqueue.AddRateLimited(workItem)
+
+				return fmt.Errorf("error processing work item '%v': %s, requeuing count %d", workItem, err.Error(), workItem.requeuingCount)
+			}
+
+			klog.Errorf("error processing work item '%v': %s, requeuing count exceeded", workItem, err.Error())
+			recordWorkItemErrorMetrics("process_work_item")
```

Comment:
I think we may want to track both total errors and errors after all retries are exhausted. Can you please emit two labels for this: `total_errors` and `errors_after_retries_exhausted`? You can emit `total_errors` inside the `if` above and `errors_after_retries_exhausted` outside it.

Reply:
do you mean to count the number of total errors inside the if with a counter and the number of errors outside the if?

Reply:
I think I understand it now but pls correct me if I was wrong, it is simply to add `recordWorkItemErrorMetrics(TotalErrors)` in the if and `recordWorkItemErrorMetrics(ErrorsAfterRetriesExhausted)` outside the if so that we can aggregate it later.